cirq/protocols/json.py

```diff
@@ -0,0 +1,234 @@
+
+    During deserialization, a class must be able to be resolved (see
+    the docstring for `read_json`) and must be able to be (re-)constructed
+    from the serialized parameters. If the type defines a classmethod
+    `_from_json_dict_`, that will be called. Otherwise, the `cirq_type` key
+    will be popped from the dictionary and used as kwargs to the type's
+    constructor.
+    """
+
+    def _json_dict_(self) -> Union[None, NotImplementedType, Dict[Any, Any]]:
+        pass
+
+
+def to_json_dict(obj, attribute_names, namespace=None):
+    if namespace is not None:
+        prefix = '{}.'.format(namespace)
+    else:
+        prefix = ''
+
+    d = {'cirq_type': prefix + obj.__class__.__name__}
+    for attr_name in attribute_names:
+        d[attr_name] = getattr(obj, attr_name)
+    return d
+
+
+class CirqEncoder(json.JSONEncoder):
+    """Extend json.JSONEncoder to support Cirq objects.
+
+    This supports custom serialization. For details, see the documentation
+    for the SupportsJSON protocol.
+
+    In addition to serializing objects that implement the SupportsJSON
+    protocol, this encoder deals with common, basic types:
+
+     - Python complex numbers get saved as a dictionary keyed by 'real'
+       and 'imag'.
+     - Numpy ndarrays are converted to lists to use the json module's
+       built-in support for lists.
+     - Preliminary support for Sympy objects. Currently only sympy.Symbol.
+       See https://github.com/quantumlib/Cirq/issues/2014
+    """
+
+    def default(self, o):
+        if hasattr(o, '_json_dict_'):
+            return o._json_dict_()
+        if isinstance(o, complex):
+            return {
+                'cirq_type': 'complex',
+                'real': o.real,
+                'imag': o.imag,
+            }
+        if isinstance(o, np.ndarray):
+            return o.tolist()
+
+        # TODO: More support for sympy
+        #       https://github.com/quantumlib/Cirq/issues/2014
+        if isinstance(o, sympy.Symbol):
+            return to_json_dict(o, ['name'], namespace='sympy')
+        return super().default(o)  # coverage: ignore
+
+
+def _cirq_object_hook(d, resolvers: List[Callable[[str], Union[None, Type]]]):
+    if 'cirq_type' in d:
+        for resolver in resolvers:
+            cls = resolver(d['cirq_type'])
+            if cls is not None:
+                break
+        else:
+            raise ValueError("Could not resolve type '{}' "
+                             "during deserialization".format(d['cirq_type']))
+
+        if hasattr(cls, '_from_json_dict_'):
+            return cls._from_json_dict_(**d)
+
+        del d['cirq_type']
+        return cls(**d)
+
+    return d
+
+
+def to_json(obj: Any, file, *, indent=2, cls=CirqEncoder):
+    if isinstance(file, str):
+        with open(file, 'w') as actually_a_file:
+            return json.dump(obj, actually_a_file, indent=indent, cls=cls)
+
+    return json.dump(obj, file, indent=indent, cls=cls)
+
+
+def read_json(
```

Comment:
This method name is not very clear. I would not expect `cirq.read_json` to return a circuit object just from reading the name. Maybe `read_from_json`?

Reply:
it's like "read (a) book" :)

again, this is from pandas

Reply:
Okay